### ISSUE TITLE
[firrtl] Improve Initialization Error Messages

### ISSUE THREAD

**Issue by seldridge** [llvm/circt#4087]
Currently, the errors generated when something is undriven are not awesome. If there is an associated FIRRTL-provided source locator, then the message is terse to the point of not being useful. If there is no source locator (or the source locator is ignored), the input FIRRTL text is then shown, but without enough context (e.g., what module it is) to fix it.

Consider:

```
circuit Foo:
  module Foo:

    wire w: UInt<1> @[herpderp 42:42]
```

Running this with `firtool`, this produces:

```
herpderp:42:42: error: sink "w" not fully initialized
herpderp:42:42: note: see current operation: %0 = "firrtl.wire"() {annotations = [], name = "w", nameKind = #firrtl<name_kind droppable_name>} : () -> !firrtl.uint<1>
```

Running with `firtool -ignore-fir-locators`:

```
﻿Foo.fir:4:5: error: sink "w" not fully initialized
    wire w: UInt<1> @[herpderp 42:42]
    ^
Foo.fir:4:5: note: see current operation: %0 = "firrtl.wire"() {annotations = [], name = "w", nameKind = #firrtl<name_kind droppable_name>} : () -> !firrtl.uint<1>
```

In a perfect world, we would show the line of Chisel that caused the issue, but we don't have that. Just printing a diagnostic that tells you what module it is in would likely be sufficient.

Ideas welcome on what would be better approach to showing this information.

h/t @azidar

**Comment by seldridge** [comment]
I was thinking about this more and there are several options: 

1. We could fuse the locations here. I'm not sure what this would do to the error message. (It would likely still be bad.)
2. We could make the default behavior to drop  the `@herpderp` locator. This is still "visible" due to the diagnostic that prints the line. (This is probably the best option.)
3. We could, when encountering a diagnostic, try to find the `@herpderp` file, pull it into a source manager, and then print out the line associated with it. I.e., lazily read the Scala file and show the line.

I think (3) is where we probably want to go eventually. This will require some changes on the Chisel side to show paths relative to some root in order disambiguate `Queue.scala` from `Queue.scala`. I'm somewhat nervous about the sheer number of source files, hence why I'm suggesting to lazily read these after the fact. If it is relatively free to add files to a SourceManager then this is, perhaps, not a problem.

For a `sifive/chisel-circt` like environment, I expect we could make this experience entirely seamless. For split Chisel/FIRRTL compiles, it would require an include root that can be used to resolve these files.

To fill out the above example, the original Chisel code is something like:

```scala
class Foo extends RawModule {
  val w = Wire(Bool())
}
```

**Comment by seldridge** [comment]
You are totally correct, @youngar. This _already_ works. Chisel may just need to emit more path information to do this correctly. Consider this example Chisel derived from this old issue (https://github.com/chipsalliance/firrtl/issues/1598):

```scala
//> using scala 2.13
//> using lib "edu.berkeley.cs::chisel3::3.5.4"
//> using lib "com.sifive::chisel-circt::0.6.0"

import chisel3._
import chisel3.stage.ChiselGeneratorAnnotation

class Bar extends MultiIOModule {
  val out = Wire(SInt())
  out := DontCare
}

object Main extends App {
  (new chisel3.stage.ChiselStage).execute(Array("--no-run-firrtl"), Seq(ChiselGeneratorAnnotation(() => new Bar)))
  (circt.stage.ChiselStage).emitSystemVerilog(new Bar)
}
```

When you run this, you get line/column with a carat pointing at the declaration. 🤯 

```
# scala-cli Bar.scala
Compiling project (Scala 2.13.8, JVM)
Warning: 1 deprecation (since Chisel 3.5); re-run with -deprecation for details
Compiled project (Scala 2.13.8, JVM)
Running CIRCT: 'firtool -format=fir -warn-on-unprocessed-annotations -verify-each=false -disable-infer-rw -dedup -annotation-file Bar.anno.json < $input'
------------------------------------------------------------------------------
Error: firtool failed.
ExitCode:
1
STDOUT:

STDERR:
Bar.scala:9:17: error: uninferred width: wire "out" is unconstrained
  val out = Wire(SInt())
                ^

------------------------------------------------------------------------------
Exception in thread "main" firrtl.options.StageError: 
	at circt.stage.phases.CIRCT.transform(CIRCT.scala:179)
	at circt.stage.phases.CIRCT.transform(CIRCT.scala:45)
	at firrtl.options.DependencyManager.$anonfun$transform$5(DependencyManager.scala:280)
	at firrtl.Utils$.time(Utils.scala:181)
	at firrtl.options.DependencyManager.$anonfun$transform$3(DependencyManager.scala:280)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:169)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:165)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at firrtl.options.DependencyManager.transform(DependencyManager.scala:269)
	at firrtl.options.DependencyManager.transform$(DependencyManager.scala:255)
	at firrtl.options.PhaseManager.transform(DependencyManager.scala:443)
	at circt.stage.ChiselStage$.emitSystemVerilog(ChiselStage.scala:102)
	at Main$.delayedEndpoint$Main$1(Bar.scala:15)
	at Main$delayedInit$body.apply(Bar.scala:13)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at scala.Function0.apply$mcV$sp$(Function0.scala:39)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:17)
	at scala.App.$anonfun$main$1(App.scala:76)
	at scala.App.$anonfun$main$1$adapted(App.scala:76)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:926)
	at scala.App.main(App.scala:76)
	at scala.App.main$(App.scala:74)
	at Main$.main(Bar.scala:13)
	at Main.main(Bar.scala)
```

**Comment by dtzSiFive** [comment]
We can still improve our initialization error messages, but at least the caret-into-sources is handled by #4539/#4576 (and modern Chisel emitting more source locators probably helps too).

One point that's been brought up is that it can be difficult for users to map the mentioned lowered names (`foo.bar` -> `foo_bar`) to their origins, especially when underscores are already in the names: is `foo_my_bar` originally `foo.my.bar` or `foo.my_bar`?

This might be why FIRRTL spec previously specified the use of `$` in lowered names (making this no longer ambiguous, `foo$my$bar` or `foo$my_bar`--even if `$` isn't disallowed in identifiers/field names it's for sure less common for user names at least :wink:), not sure of the history there.

**Comment by seldridge** [comment]
I remember there being some friction with use of `$`. This is legal for Verilog identifiers and used to be legal for FIRRTL identifiers (as you've noticed in the spec). However, there were anecdotal reports of problems with use of `$` for buggy Verilog parsers or just general confusion of `$`. I eventually just changed all of this to use `_` instead of `$`. See: https://github.com/chipsalliance/firrtl/issues/729, https://github.com/chipsalliance/firrtl/pull/909

One approach towards improving this _specific_ error message would be to see about moving `ExpandWhens` before `LowerTypes`. (And once aggregate preservation is on, `LowerTypes` is mostly a no-op except for spec-mandated places where types have to be lowered.)

**Comment by dtzSiFive** [comment]
FWIW, PoC exploring options re:fusing and attaching to diagnostics to get both: https://github.com/llvm/circt/pull/4671 .
